### ISSUE TITLE
Transfer learning

### ISSUE THREAD

**Issue by lgvaz** [PyTorchLightning/pytorch-lightning#1991]
# The problem

The standard flow for transfer learning is as follows:

* Start from a pretrained model and create a new head
* Freeze all layers but the head
* Train
* Unfreeze all (or some) layers
* Train (with differential learning rates)

Let's call each "freeze/unfreeze -> train" step a phase.

What can change between phases?
* The `lr_scheduler` should restart
* New parameter groups can be added/removed to the `optimizer` and to the `lr_scheduler` (We don't really add a parameter group to the `lr_scheduler`, but we should pass a list of learning rates, one for each group (this is called differential learning rates))

# Parameter groups  

I've implemented a first idea on how to implement this as a ModuleMixin [here](https://github.com/lgvaz/mantisshrimp/blob/master/mantisshrimp/models/parameters_splits_module_mixin.py), the idea is for the user to override `model_splits`, which returns the parameter groups of the model.  

You can see how this gets used for differential learning rates [here](https://github.com/lgvaz/mantisshrimp/blob/master/mantisshrimp/models/prepare_optimizer_module_mixin.py#L24)

---
# Changing phases
There are two main ways of changing phases, but currently none of them are smooth.

## Use multiple trainers
Use a separate trainer for each phase, always resuming the last state by using `resume_from_checkpoint`.  
A scratch notebook for this idea is [here](https://colab.research.google.com/drive/13MGi7Kot3VVOqSpEtmGi5kEGJQ_OQBe3?usp=sharing)  
The problem is that trainer also resumes the `lr_scheduler` state, but we need it to start from zero

## Implement the logic in `on_epoch_start`
[Idea notebook](https://colab.research.google.com/drive/1OrpPnr5X-x9yvlZd1zjTx6wEZ-mmSOEL?usp=sharing)
* Check the `current_epoch` and change phases when needed
* Might need to add/remove parameter groups from the optimizer/lr_scheduler
* Changing the scheduler is causing problems when used together with `LeraningRateLogger`

**Comment by lgvaz** [comment]
Currently, if you look at `get_optimizer_param_groups` [here](https://github.com/lgvaz/mantisshrimp/blob/master/mantisshrimp/models/prepare_optimizer_module_mixin.py#L24) you can see that I'm only adding the `trainable_params` to the optmizer, but we could also add all parameters to it, if `requires_grad=False` for a parameter it will be ignored any ways.

Always adding all params might eliminate the problem of having to add/remove new parameters to the optimizer.

**Comment by lgvaz** [comment]
I create a notebook with everything kinda of working [here](https://colab.research.google.com/drive/1OrpPnr5X-x9yvlZd1zjTx6wEZ-mmSOEL?usp=sharing), using the second method described (`on_epoch_start`)

I also updated the OP with the link
